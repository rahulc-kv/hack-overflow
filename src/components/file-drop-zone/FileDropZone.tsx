import React, { FC, useCallback } from 'react';
import { useDropzone, FileRejection } from 'react-dropzone';

import { StringToTSX } from '@components/stringToTsx/StringToTsx';
import { ExportDynamicIcon, FileIcon } from '@assets/icons';
import { translate } from '@utils/locale';
import {
  SUPPORTED_FILE_EXTENSIONS,
  SUPPORTED_FILE_FORMATS
} from '@constants/common';
import { FileDropZoneProps } from './types';
import COLORS from '@constants/colors';

const FileDropZone: FC<FileDropZoneProps> = props => {
  const {
    maxFileCount,
    handleFileSelect,
    isDisabled = false,
    currentUploadedFileCount,
    exchangeCode = 'DEFAULT',
    fileFormats = SUPPORTED_FILE_EXTENSIONS.DEFAULT,
    uploaderSize = 'large'
  } = props;

  const handleOnDrop = useCallback(
    (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
      const formattedRejectedFiles = rejectedFiles.map(
        fileData => fileData.file
      );
      // total number of selected file will be acceptedFiles.length + currentUploadedFileCount
      /*
         if the total number of files is greater than maxFileCount, then we will reject the
         remaining files after accepting up to maxFileCount
      */
      if (acceptedFiles.length + currentUploadedFileCount > maxFileCount) {
        /* 
          We will accept only up to "maxFileCount". If the count of user-picked files is greater
          than maxFileCount, then we will take the first n items and the rest of the items 
          will consider as rejected files
        */
        const slicingIndex = maxFileCount - currentUploadedFileCount;
        /*
          first param is acceptedFiles and the second one is rejectedFiles. In this case 
          the rejectedFiles will be a concatenated array of 'formattedRejectedFiles' 
          and the remaining files after accepting the allowed number of files from 
          the list picked by the user 
        */
        handleFileSelect([...acceptedFiles].slice(0, slicingIndex), [
          ...[...acceptedFiles].slice(slicingIndex),
          ...formattedRejectedFiles
        ]);
        return;
      }
      handleFileSelect(acceptedFiles, formattedRejectedFiles);
    },
    [currentUploadedFileCount, maxFileCount]
  );

  const { getRootProps, getInputProps } = useDropzone({
    accept:
      SUPPORTED_FILE_FORMATS[exchangeCode] || SUPPORTED_FILE_FORMATS.DEFAULT,
    multiple: true,
    onDrop: handleOnDrop,
    disabled: isDisabled
  });

  return (
    <>
      {uploaderSize === 'large' ? (
        <div
          className={`flex flex-col justify-center items-center px-2 py-[18px] text-center rounded-md border-[1px]
        ${
          isDisabled
            ? 'border-greyCloud cursor-not-allowed'
            : 'border-primaryColor cursor-pointer'
        } border-dashed`}
          {...getRootProps()}>
          <FileIcon
            fill={isDisabled ? COLORS.GREY_CLOUD : COLORS.PRIMARY_COLOR}
          />
          <input data-testid="dropInput" {...getInputProps()} />
          <p className="mt-[6px] text-sm font-normal text-davyGrey">
            {
              <StringToTSX
                domString={translate('sourcePage.dropFile', {
                  color: isDisabled ? COLORS.SLATE_GREY : COLORS.PRIMARY_COLOR
                })}
              />
            }
          </p>
          <p className="text-sm font-normal text-slateGrey ">{fileFormats}</p>
        </div>
      ) : (
        <div
          className={`flex flex-row justify-start items-center px-4 py-[6px] text-center rounded-md border-[1px]
      ${
        isDisabled
          ? 'border-greyCloud cursor-not-allowed'
          : 'border-primaryColor cursor-pointer'
      } border-dashed`}
          {...getRootProps()}>
          <ExportDynamicIcon
            fill={isDisabled ? COLORS.GREY_CLOUD : COLORS.BLACK_GREEN}
          />
          <p className="ml-[11px] text-sm font-normal text-davyGrey">
            {
              <StringToTSX
                domString={translate('sourcePage.dropFile', {
                  color: isDisabled ? COLORS.SLATE_GREY : COLORS.PRIMARY_COLOR
                })}
              />
            }
          </p>
          <input data-testid="dropInput" {...getInputProps()} />
        </div>
      )}
    </>
  );
};

export default FileDropZone;
