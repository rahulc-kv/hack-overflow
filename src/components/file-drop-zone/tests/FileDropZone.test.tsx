import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import FileDropZone from '@components/file-drop-zone/FileDropZone';

const handleFileSelect = jest.fn();

test('FileDropZone drop', async () => {
  const component = render(
    <FileDropZone
      maxFileCount={5}
      handleFileSelect={handleFileSelect}
      isDisabled={true}
      currentUploadedFileCount={3}
      exchangeCode={'WRX'}
      fileFormats={'.xlsx'}
    />
  );
  expect(component).toMatchSnapshot();
});

test('File Dropped', () => {
  render(
    <FileDropZone
      maxFileCount={5}
      handleFileSelect={handleFileSelect}
      isDisabled={true}
      currentUploadedFileCount={3}
      exchangeCode={'WRX'}
      fileFormats={'.xlsx'}
    />
  );
  const dropInput = screen.getByTestId('dropInput');
  fireEvent.drop(dropInput, {
    DataTransfer: {
      files: [new File(['file'], 'test.xlsx', { type: '' })]
    }
  });
});

//TODO: Add the rejected scenario after writing test cases for all file components.
