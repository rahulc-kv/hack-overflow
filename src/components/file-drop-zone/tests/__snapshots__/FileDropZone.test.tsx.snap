// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileDropZone drop 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col justify-center items-center px-2 py-[18px] text-center rounded-md border-[1px]
        border-greyCloud cursor-not-allowed border-dashed"
        role="presentation"
      >
        <svg
          fill="#B4B7BD"
        >
          fileIcon.svg
        </svg>
        <input
          accept="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,.xlsx"
          data-testid="dropInput"
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <p
          class="mt-[6px] text-sm font-normal text-davyGrey"
        >
          Drop or 
          <span
            style="color: rgb(126, 129, 141);"
          >
             Browse 
          </span>
           your file
        </p>
        <p
          class="text-sm font-normal text-slateGrey "
        >
          .xlsx
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col justify-center items-center px-2 py-[18px] text-center rounded-md border-[1px]
        border-greyCloud cursor-not-allowed border-dashed"
      role="presentation"
    >
      <svg
        fill="#B4B7BD"
      >
        fileIcon.svg
      </svg>
      <input
        accept="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,.xlsx"
        data-testid="dropInput"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
      <p
        class="mt-[6px] text-sm font-normal text-davyGrey"
      >
        Drop or 
        <span
          style="color: rgb(126, 129, 141);"
        >
           Browse 
        </span>
         your file
      </p>
      <p
        class="text-sm font-normal text-slateGrey "
      >
        .xlsx
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
